### PR TITLE
[White Label] Make logo url link more tolerant in terms of format

### LINKED ISSUE #10903: [White Label] Make logo url link more tolerant in terms of format and add example
## What is the need / problem?
Setting a logo url link in the format `http://www.url.com` works as expected and opens that page in the same browser tab. :heavy_check_mark:
Setting a link in a different format, e.g. `www.url.com` or `url.com` links to `CURRENT_PAGE/url.com` which leads to a 404 page not found error. :x:
To make the link URL input a bit more tolerant we could try using the same or similar methods as we do on the enterprise's website URL on the contact tab.
Also we should add an example as a placeholder, as we do on the social and contact tabs in enterprise settings.

## Which type of users does this problem affect (and how many, if known)?
- Hub Managers (MCFEs)
- Instance Managers

## Connected wishlist and discovery discussions
Follow up of [#10586](https://github.com/openfoodfoundation/openfoodnetwork/pull/10586).

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

- Closes #10903

Makes URLs used in `LINK FOR THE LOGO USED IN SHOPFRONT` under the white label section, absolutized. 

#### What should we test?

- As a shop manager, edit the `LINK FOR THE LOGO USED IN SHOPFRONT` value. 
- Check that entering an URL without `http` or `https`, create actually an URL with a `http://` added
- Check that entering an URL starting with `http`, doesn't change the value of the URL

Therefor we do have an absolute URL.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/enterprise.rb [hunk 1]
```diff
@@ -317,6 +317,14 @@ def white_label_logo_url(name = :default)
     )
   end
 
+  def white_label_logo_link
+    return nil if self[:white_label_logo_link].blank?  
+
+    return self[:white_label_logo_link] if self[:white_label_logo_link].start_with?('http')
+
+    "http://#{self[:white_label_logo_link]}"
```
Comment (mkllnk) [inline app/models/enterprise.rb:325]: Shouldn't we use https by default? :shrug: 

Or maybe we should validate the URL on the model earlier. :thinking:

File: spec/system/admin/enterprises_spec.rb [hunk 2]
```diff
@@ -710,6 +710,8 @@
 
           context "can edit white label logo link" do
             it_behaves_like "edit link with", "https://www.openfoodnetwork.org", "https://www.openfoodnetwork.org"
+            it_behaves_like "edit link with", "www.openfoodnetwork.org", "http://www.openfoodnetwork.org"
+            it_behaves_like "edit link with", "openfoodnetwork.org", "http://openfoodnetwork.org"
```
Comment (mkllnk) [inline spec/system/admin/enterprises_spec.rb:714]: I would have tested this on the model because it's much faster.

**Review by mkllnk — APPROVED** [review]
Great, solved.

**Review by dacook — CHANGES_REQUESTED** [review]
I agree that we should try and validate this earlier to ensure the user supplies the correct value, rather than manipulating the string later.

**Comment by jibees** [comment]
> I agree that we should try and validate this earlier to ensure the user supplies the correct value, rather than manipulating the string later.

I previously did not choose this solution, in order to be backward compatible with previous values. But you're right, it's not that important. On the other hand, validation based on HTML component, still allows user to POST _bad_ data. But it's an edge case, I agree. 
Back `In Dev` to change that, thanks!

**Comment by jibees** [comment]
It's not that straightforward since we're on AngularJS form, and we need to take into account that validation by native HTML component. Will try.

**Comment by mkllnk** [comment]
Can we also add a validation to the Enterprise model?

I'm just wondering which logic browsers apply when you type in the address bar. When you type example.com then it adds the http:// automatically. It's safe to assume http or https. I don't think that we need to force users to add http://.

There's probably a regex for valid URLs. Otherwise we can validate on the Enterprise model like this:

```
uri = URI(s)
uri = URI("http://#{s}") if uri.scheme.nil?
if uri.is_a? URI::HTTP # includes https
  self.white_label_logo_link = uri.to_s
else
  errors.add(...) unless uri.is_a? URI::HTTP # includes https
end
```

We also have an indirect dependency to the [addressable gem](https://github.com/sporkmonger/addressable). We could make that explicit and use its heuristic parsing:

* https://www.rubydoc.info/gems/addressable/Addressable/URI#heuristic_parse-class_method

File: app/models/enterprise.rb [hunk 3]
```diff
@@ -462,6 +455,17 @@ def devise_mailer
 
   private
 
+  def validate_white_label_logo_link
+    return if white_label_logo.blank?
```
Comment (dacook) [inline app/models/enterprise.rb:459]: I was going to suggest this dependency isn't necessary, but then remembered the link input field is hidden when the checkbox is unchecked. 
So it's not fair on the user to validate something they can't fix. (Eg if we removed this check, then the validation rules change in the future, an old hidden value could cause an error that the user can't see).

So all good 👍

File: app/models/enterprise.rb [hunk 4]
```diff
@@ -462,6 +455,17 @@ def devise_mailer
...
+
+    return if white_label_logo_link.blank?
+
+    uri = URI(white_label_logo_link)
+    self.white_label_logo_link = "http://#{white_label_logo_link}" if uri.scheme.nil?
```
Comment (dacook) [inline app/models/enterprise.rb:464]: I would have expected `https` to be the default, but it doesn't really matter. If it matters, the user can update it!

**Review by dacook — APPROVED** [review]
Thanks for updating, great specs!

File: app/models/enterprise.rb [hunk 5]
```diff
@@ -462,6 +455,17 @@ def devise_mailer
 
   private
 
+  def validate_white_label_logo_link
+    return if white_label_logo.blank?
+
+    return if white_label_logo_link.blank?
+
+    uri = URI(white_label_logo_link)
```
Comment (mkllnk) [inline app/models/enterprise.rb:463]: Nice. We could call `strip` on the input in case someone copies whitespace but it's not necessary.

**Review by mkllnk — APPROVED** [review]

**Comment by mkllnk** [comment]
I decided that adding `strip` is small enough to quickly add and still progress to testing.

**Comment by jibees** [comment]
> I decided that adding `strip` is small enough to quickly add and still progress to testing.

I agree. Thanks for updating my PR!

File: spec/models/enterprise_spec.rb [hunk 6]
```diff
@@ -332,6 +332,12 @@
         expect(e.white_label_logo_link).to eq "http://www.example.com"
       end
 
+      it "ignores whitespace around the URL form copying and pasting" do
+        e = build(:enterprise, white_label_logo_link: ' www.example.com ')
+        expect(e).to be_valid
+        expect(e.white_label_logo_link).to eq "http://www.example.com"
+      end
```
Comment (jibees) [inline spec/models/enterprise_spec.rb:339]: Thanks @mkllnk !!

**Comment by filipefurtad0** [comment]
Hey @jibees ,


As a shop manager, editing the LINK FOR THE LOGO USED IN SHOPFRONT value:

- Checked that entering an URL without `http` or `https`, create actually an URL with a `http://` added

`github.com/` -> `https://github.com/` :heavy_check_mark: 

- Check that entering an URL starting with http, doesn't change the value of the URL

`https://github.com/` -> `https://github.com/` :heavy_check_mark: 


The PR also introduces a validation, right? So, I've tried a test case "between the cracks", an invalid URL:

`://github.com/ ` -> validation error :heavy_check_mark: 
but a missing translation, too:
image

This looks good to go, I'll open an issue for the missing translation.

**Pull request merged**